In [8]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [10]:
%sql postgresql://appdev@data:5432/appdev

'Connected: appdev@appdev'

Ex 1
The table contains three indexes: 
idx_17102_primary 
Using a b-tree for this index helps optimizing searching. 
This index has a size of 16kB. 

idx_17102_url
As said a b-tree is god for searching. In this case its great because it can be sorted alphabetically. 

idx_17102_url
This is an index of type GiST. 
GiST is good to use for two-dimensional geometric data types. 
As we can see, position is of type point, which is a data type containing (x,y). 
GiST indexes can also be used for optimizing "nearest-neighbor" searches, which could be relevant in this case. 
This index has a size of 8192 byte. 

In [56]:
%sql SELECT races.date, surname, dob, milliseconds, races.name racename, circuits.name FROM results JOIN races USING(raceid) JOIN drivers USING (driverid)JOIN circuits USING(circuitid) WHERE dob > '1980-01-01' AND rank = 1;

162 rows affected.


[(datetime.date(2008, 3, 16), 'Kovalainen', datetime.date(1981, 10, 19), 5708630, 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2008, 4, 6), 'Kovalainen', datetime.date(1981, 10, 19), 5493759, 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2008, 8, 24), 'Massa', datetime.date(1981, 4, 25), 5732339, 'European Grand Prix', 'Valencia Street Circuit'),
 (datetime.date(2008, 10, 12), 'Massa', datetime.date(1981, 4, 25), 5468050, 'Japanese Grand Prix', 'Fuji Speedway'),
 (datetime.date(2008, 10, 19), 'Hamilton', datetime.date(1985, 1, 7), 5517403, 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2008, 11, 2), 'Massa', datetime.date(1981, 4, 25), 5651435, 'Brazilian Grand Prix', 'Autódromo José Carlos Pace'),
 (datetime.date(2007, 4, 8), 'Hamilton', datetime.date(1985, 1, 7), 5552487, 'Malaysian Grand Prix', 'Sepang International Circuit'),
 (datetime.date(2007, 4, 15), 'Massa', datetime.date(1981, 4, 25), 5607515, 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2007, 5, 13), 'Massa', datetime.date(1981, 4, 25), 5496230, 'Spanish Grand Prix', 'Circuit de Barcelona-Catalunya'),
 (datetime.date(2007, 5, 27), 'Alonso', datetime.date(1981, 7, 29), 6029329, 'Monaco Grand Prix', 'Circuit de Monaco'),
 (datetime.date(2007, 6, 10), 'Alonso', datetime.date(1981, 7, 29), 6273228, 'Canadian Grand Prix', 'Circuit Gilles Villeneuve'),
 (datetime.date(2007, 7, 1), 'Massa', datetime.date(1981, 4, 25), 5456614, 'French Grand Prix', 'Circuit de Nevers Magny-Cours'),
 (datetime.date(2007, 7, 22), 'Massa', datetime.date(1981, 4, 25), 7594513, 'European Grand Prix', 'Nürburgring'),
 (datetime.date(2007, 9, 9), 'Alonso', datetime.date(1981, 7, 29), 4717806, 'Italian Grand Prix', 'Autodromo Nazionale di Monza'),
 (datetime.date(2007, 9, 16), 'Massa', datetime.date(1981, 4, 25), 4843761, 'Belgian Grand Prix', 'Circuit de Spa-Francorchamps'),
 (datetime.date(2007, 9, 30), 'Hamilton', datetime.date(1985, 1, 7), 7234579, 'Japanese Grand Prix', 'Fuji Speedway'),
 (datetime.date(2007, 10, 7), 'Massa', datetime.date(1981, 4, 25), 5891286, 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2006, 3, 12), 'Rosberg', datetime.date(1985, 6, 27), 5449248, 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2006, 3, 19), 'Alonso', datetime.date(1981, 7, 29), 5445114, 'Malaysian Grand Prix', 'Sepang International Circuit'),
 (datetime.date(2006, 4, 23), 'Alonso', datetime.date(1981, 7, 29), 5468582, 'San Marino Grand Prix', 'Autodromo Enzo e Dino Ferrari'),
 (datetime.date(2006, 5, 14), 'Massa', datetime.date(1981, 4, 25), 5211612, 'Spanish Grand Prix', 'Circuit de Barcelona-Catalunya'),
 (datetime.date(2006, 6, 11), 'Alonso', datetime.date(1981, 7, 29), 5151927, 'British Grand Prix', 'Silverstone Circuit'),
 (datetime.date(2006, 8, 6), 'Massa', datetime.date(1981, 4, 25), None, 'Hungarian Grand Prix', 'Hungaroring'),
 (datetime.date(2006, 10, 1), 'Alonso', datetime.date(1981, 7, 29), 5855868, 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2006, 10, 8), 'Alonso', datetime.date(1981, 7, 29), 5032413, 'Japanese Grand Prix', 'Suzuka Circuit'),
 (datetime.date(2005, 3, 6), 'Alonso', datetime.date(1981, 7, 29), 5064048, 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2005, 5, 29), 'Alonso', datetime.date(1981, 7, 29), 5506648, 'European Grand Prix', 'Nürburgring'),
 (datetime.date(2009, 3, 29), 'Rosberg', datetime.date(1985, 6, 27), 5661506, 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2009, 4, 5), 'Button', datetime.date(1980, 1, 19), 4252092, 'Malaysian Grand Prix', 'Sepang International Circuit'),
 (datetime.date(2009, 5, 24), 'Massa', datetime.date(1981, 4, 25), 6059392, 'Monaco Grand Prix', 'Circuit de Monaco'),
 (datetime.date(2009, 6, 7), 'Button', datetime.date(1980, 1, 19), 5184848, 'Turkish Grand Prix', 'Istanbul Park'),
 (datetime.date(20

In [50]:
%sql EXPLAIN ANALYZE SELECT races.date, surname, dob, milliseconds, races.name racename, circuits.name FROM results JOIN races USING(raceid) JOIN drivers USING (driverid)JOIN circuits USING(circuitid) WHERE dob > '1980-01-01' AND rank = 1;

18 rows affected.


[('Nested Loop  (cost=25.68..752.35 rows=18 width=62) (actual time=0.213..12.467 rows=162 loops=1)',),
 ('  ->  Nested Loop  (cost=25.54..749.18 rows=18 width=50) (actual time=0.207..8.203 rows=162 loops=1)',),
 ('        ->  Hash Join  (cost=25.26..737.47 rows=18 width=27) (actual time=0.197..7.764 rows=162 loops=1)',),
 ('              Hash Cond: (results.driverid = drivers.driverid)',),
 ('              ->  Seq Scan on results  (cost=0.00..708.96 rows=257 width=24) (actual time=0.020..7.497 rows=257 loops=1)',),
 ('                    Filter: (rank = 1)',),
 ('                    Rows Removed by Filter: 23420',),
 ('              ->  Hash  (cost=24.51..24.51 rows=60 width=19) (actual time=0.167..0.167 rows=61 loops=1)',),
 ('                    Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                    ->  Seq Scan on drivers  (cost=0.00..24.51 rows=60 width=19) (actual time=0.010..0.151 rows=61 loops=1)',),
 ("                          Filter: (dob > '1980-01-01'::date)",),
 ('                          Rows Removed by Filter: 780',),
 ('        ->  Index Scan using idx_17199_primary on races  (cost=0.28..0.65 rows=1 width=39) (actual time=0.002..0.002 rows=1 loops=162)',),
 ('              Index Cond: (raceid = results.raceid)',),
 ('  ->  Index Scan using idx_17102_primary on circuits  (cost=0.14..0.18 rows=1 width=28) (actual time=0.001..0.001 rows=1 loops=162)',),
 ('        Index Cond: (circuitid = races.circuitid)',),
 ('Planning time: 5.143 ms',),
 ('Execution time: 12.542 ms',)]

Ex 2: 
I think it perform 9 calls.
Execution time is 12.542 ms. 
It explains every step and how long it takes to execute. 
In case of filter it also display how many rows it removes with this filter. 
For example the rank = 1 results in 23420 rows being removed. 


In [52]:
%sql CREATE MATERIALIZED VIEW race_winners_caches AS SELECT races.date, surname, dob, milliseconds, races.name racename, circuits.name FROM results JOIN races USING(raceid) JOIN drivers USING (driverid)JOIN circuits USING(circuitid) WHERE dob > '1980-01-01' AND rank = 1;

162 rows affected.


[]

In [55]:
%sql EXPLAIN ANALYZE SELECT * from race_winners_caches

3 rows affected.


[('Seq Scan on race_winners_caches  (cost=0.00..4.62 rows=162 width=66) (actual time=0.013..0.030 rows=162 loops=1)',),
 ('Planning time: 0.156 ms',),
 ('Execution time: 0.082 ms',)]

With MATERIALIZED VIEW we make a view which wil be cached. 
This is why the execution time is so short- because it have been cached, and it won't have to execute the whole query again. 